do not assume DTM version unless dtmcontrol is read successfully

aap-sc opened the pull request:
Change-Id: I02906dbbe04778623a51774bc21fa45b55d12881

aap-sc committed:
do not assume DTM version unless dtmcontrol is read successfully

Change-Id: I5f2003b7ac5ce87af6ca9a4fcb46140682a8cfdf
Signed-off-by: Parshintsev Anatoly <anatoly.parshintsev@syntacore.com>

aap-sc commented:
@timsifive I know that changes to riscv-0.11 are discouraged, however in this case I believe it's better to add this fix. There were a couple of serious bugs:

1. dtmcontrol_scan could return an error code as a scanned value.
2. in case when `dtmcontrol` was read as "0" we had an implicit fallback to v0.11 that caused register cache to be populated. This could result in a weird situations like:
```
XLEN=64, misa=0x14112f
```

timsifive commented on src/target/riscv/riscv-011.c:
```diff
@@ -273,7 +273,7 @@ static uint16_t dram_address(unsigned int index)
 }
 
-static uint32_t dtmcontrol_scan(struct target *target, uint32_t out)
+static uint32_t dtmcontrol_scan(struct target *target, uint32_t out, uint32_t *in_ptr)
```
Here and elsewhere, if you're now returning ERROR_ values, the type should be int.

aap-sc replied:
my bad. will fix.

aap-sc replied:
addressed.

timsifive commented on src/target/riscv/riscv.c:
```diff
@@ -411,11 +414,13 @@ static struct target_type *get_target_type(struct target *target)
 			return &riscv013_target;
 		default:
+			// TODO: once we have proper support for non-examined targets
+			// we should have an assert here
```
I think OpenOCD still wants /* */ comments (although the enforcement disappeared from the checkpatch script at some point).

aap-sc replied:
addressed

timsifive commented on src/target/riscv/riscv.c:
```diff
@@ -1545,8 +1555,14 @@ static int riscv_examine(struct target *target)
+		return ERROR_FAIL;
+	}
 	LOG_TARGET_DEBUG(target, "dtmcontrol=0x%x", dtmcontrol);
+	// FIXME: if examine fails we may need to restore the original dtm_version
```
That makes sense. Can you add that change? Shouldn't be much more complicated.

aap-sc replied:
will try... though I'm not sure that we need to fix it in this patch. `examine` has it's share of issues:

for example, it halts the target, but if something goes wrong - the target is not resumed. The fix for such situations may be complicated. But having said that - I'll try to address at least `dtm_version` stuff

aap-sc replied:
addressed..

Apparently my whining about examine implementation is not relevant to `riscv.c`

timsifive reviewed (changes requested):
How did you test the riscv-011 changes?

aap-sc commented:
> How did you test the riscv-011 changes?

Regarding this...
-  I haven't tested on v0.11 since I don't have such HW, unfortunately. As far as v0.11 goes I've only tried to make sure that the change is reasonable. 
- In case of v0.13 there was extensive testing: the whole riscv_tests and additional tests on fpga

TommyMurphyTM1234 commented:
> > How did you test the riscv-011 changes?
> 
> Regarding this...
> 
> * I haven't tested on v0.11 since I don't have such HW, unfortunately. As far as v0.11 goes I've only tried to make sure that the change is reasonable by visually inspecting the code.

If it helps, I can run the tests against a HiFive1 (Rev A I think) in the next couple of days?

aap-sc commented:
> If it helps, I can run the tests against a HiFive1 (Rev A I think) in the next couple of days?

Oh, it would be great! If you have the bandwidth - please, do. Though I'll need to address few comments by @timsifive

aap-sc commented:
> I take it that this PR still needs some changes before it makes sense for me to test against the HiFive1?

I've already addressed all concerns raised by @timsifive  . Maybe we should wait for Tim to confirm that this is good enough.

JanMatCodasip commented on src/target/riscv/riscv-013.c:
```diff
@@ -1845,17 +1847,19 @@ static int examine(struct target *target)
-	uint32_t dtmcontrol = dtmcontrol_scan(target, 0);
+	uint32_t dtmcontrol;
+	if (dtmcontrol_scan(target, 0, &dtmcontrol) != ERROR_OK || dtmcontrol == 0) {
+		LOG_TARGET_ERROR(target, "could not scan dtmcontrol. Check JTAG connectivity/board power.");
```
Just nit-picking: 

```suggestion
		LOG_TARGET_ERROR(target, "Could not scan dtmcontrol. Check JTAG connectivity/board power.");
```

aap-sc replied:
addressed

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -389,14 +390,16 @@ static uint32_t dtmcontrol_scan(struct target *target, uint32_t out)
 	int retval = jtag_execute_queue();
 	if (retval != ERROR_OK) {
-		LOG_ERROR("failed jtag scan: %d", retval);
+		LOG_TARGET_ERROR(target, "failed dtmcontrol scan, error code = %d", retval);
```
```suggestion
		LOG_TARGET_ERROR(target, "dtmcontrol scan failed, error code = %d", retval);
```

aap-sc replied:
addressed

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -1542,25 +1552,40 @@ static int riscv_examine(struct target *target)
 	struct target_type *tt = get_target_type(target);
 	if (!tt)
-		return ERROR_FAIL;
+		goto report_examination_failure;
```
The goto label can be shorter, if you like. For example:

```suggestion
		goto examine_fail;
```

aap-sc replied:
addressed

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -1542,25 +1552,40 @@ static int riscv_examine(struct target *target)
-	int result = tt->init_target(info->cmd_ctx, target);
-	if (result != ERROR_OK)
-		return result;
+	examination_status = tt->init_target(info->cmd_ctx, target);
```
This can be shortened to:

```suggestion
	examine_status = tt->init_target(info->cmd_ctx, target);
```

It becomes easier to read - 3 syllables instead of 5.

aap-sc replied:
addressed

JanMatCodasip commented on src/target/riscv/riscv.c:
```diff
@@ -1542,25 +1552,40 @@ static int riscv_examine(struct target *target)
 
-	return tt->examine(target);
+report_examination_failure:
+	info->dtm_version = DTM_DTMCS_VERSION_UNKNOWN;
```
If examination of the DTM succeeds (dtmcontrol_scan()) but the later examination of the target fails, I believe we should keep the value of `info->dtm_version` because the DTM examination actually passed.

In that case, we should still be able to issue e.g. low-level DMI commands (`riscv dmi_read`, `riscv dmi_write`). I am not sure if that works today, but I believe we should not close the door on that.

timsifive replied:
I would prefer to reset dtm_version if examine() fails. If examine() fails, it might be because the target is in some wonky state, held in reset, whatever, and the dtm version itself might be suspect. (Obviously this is not supposed to be that way, but hardware isn't always the way it's supposed to be, especially during initialization.) The added cost of detecting it again is very low

JanMatCodasip replied:
Ok, good point, agreed.

JanMatCodasip reviewed (commented):
I like this fix/improvement and the increased level of robustness it brings.

I have only few minor comments, otherwise it looks fine.

JanMatCodasip reviewed (approved):
The changes look good to me. 

When testing on HiFive01 Rev A01 with [known fails](https://github.com/riscv/riscv-openocd/issues/869) skipped:

- On HiFive1.py target, the testing passed without any issues.
- On HiFive1-flash.py, there were 8 exceptions, occurring both with and without this change. I haven't investigated these further.

```
:::::::::::::::::::::::::::[ ran 72 tests in 465s ]:::::::::::::::::::::::::::
28 tests returned not_applicable
36 tests returned pass
8 tests returned exception
   DebugBreakpoint > logs/20231009-082756-HiFive1Flash-DebugBreakpoint.log
   DebugExit > logs/20231009-082844-HiFive1Flash-DebugExit.log
   DebugFunctionCall > logs/20231009-082926-HiFive1Flash-DebugFunctionCall.log
   Hwbp1 > logs/20231009-083025-HiFive1Flash-Hwbp1.log
   Hwbp2 > logs/20231009-083108-HiFive1Flash-Hwbp2.log
   Registers > logs/20231009-083240-HiFive1Flash-Registers.log
   TooManyHwbp > logs/20231009-083342-HiFive1Flash-TooManyHwbp.log
   UserInterrupt > logs/20231009-083442-HiFive1Flash-UserInterrupt.log

```

TommyMurphyTM1234 commented:
> The changes look good to me.
> 
> When testing on HiFive01 Rev A01 with [known fails](https://github.com/riscv/riscv-openocd/issues/869) skipped:
> 
> * On HiFive1.py target, the testing passed without any issues.
> * On HiFive1-flash.py, there were 8 exceptions, occurring both with and without this change. I haven't investigated these further.

If it's of any use, I plan to look at this imminently and could also look at those additional (?) exceptions?

* https://github.com/riscv-software-src/riscv-tests/issues/488

JanMatCodasip commented:
> If it's of any use, I plan to look at this imminently and could also look at those additional (?) exceptions?

Any help with the failing tests is appreciated :)

aap-sc commented:
Hi @timsifive , just wanted to double-check if there are any additional changes/fixes required for this PR?
